network/p2p/connection.go

```diff
@@ -82,32 +87,166 @@ func (conn *connection) send(message *transport.NetworkMessage) error {
 	return nil
 }
 
-// sendMessages (blocking) reads messages from its outMessages channel and sends them to the peer until the channel is closed.
-func (conn connection) sendMessages() {
-	for message := range conn.outMessages {
-		if conn.messenger.Send(message) != nil {
-			log.Logger().Warnf("Unable to send message to peer (peer=%s)", conn.Peer)
+func (conn *managedConnection) exchange(receivedMessages messageQueue) {
+	conn.mux.Lock()
+	out := conn.outMessages
+	in := conn.receiveMessages()
+	conn.mux.Unlock()
+	for {
```

Comment:
sending is preferred over reading. If you send a lot, would this overflow the in channel?

Reply:
Good point, I actually don't know how this functions.

Tested it, turns out it's some it's sort of random although the first select `case` will always have precidence:

```
func TestSelectChanPrio(t *testing.T) {
	const count = 5
	chanA := make(chan string, count)
	chanB := make(chan string, count)
	for i := 0; i < count; i++ {
		chanA <- "A"
	}
	for i := 0; i < count; i++ {
		chanB <- "B"
	}

	for {
		select {
		case item := <-chanA:
			println(item)
		case item := <-chanB:
			println(item)
		}
	}
}
```

Outputs:

```
AABABABBBA
AABABBABAB
AABAABABBB
ABABABBBAA
```

It's not 100% fair, but good enough I think?

Reply:
It behaves a bit like go routines, so no need to use go routines 👍